src/os/linux/xdg.rs

```diff
@@ -0,0 +1,409 @@
+use freedesktop_entry_parser::parse_entry;
+use freedesktop_entry_parser::Entry as FreeDesktopEntry;
+use once_cell_regex::{byte_regex, exports::regex::bytes::Regex};
+use std::{
+    env,
+    ffi::{OsStr, OsString},
+    os::unix::ffi::{OsStrExt, OsStringExt},
+    path::{Path, PathBuf},
+};
+
+// Detects which .desktop file contains the data on how to handle a given
+// mime type (like: "with which program do I open a text/rust file?")
+pub fn query_mime_entry(mime_type: &str) -> Option<PathBuf> {
+    let out = bossy::Command::impure("xdg-mime")
+        .add_args(&["query", "default", mime_type])
+        .run_and_wait_for_output()
+        .ok()?;
+    if let Ok(out_str) = out.stdout_str() {
+        log::debug!("query_mime_entry got output \"{}\"", out_str);
+        if !out_str.is_empty() {
+            return Some(out_str.trim().into());
+        }
+    }
+    None
+}
+
+// Returns the first entry on that directory whose filename is equal to target.
+//
+// This spec is what makes me believe the search is recursive:
+// https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html
+// This other one does not give that idea:
+// https://specifications.freedesktop.org/menu-spec/latest/ar01s02.html
+pub fn find_entry_in_dir(dir_path: &Path, target: &Path) -> Option<PathBuf> {
+    for entry in dir_path.read_dir().ok()? {
+        if let Ok(entry) = entry {
+            // If it is a file with that same _filename_ (not full path)
+            if entry.path().is_file() && entry.file_name() == target {
+                return Some(entry.path().into());
+            } else if entry.path().is_dir() {
+                // I think if there are any dirs on that directory we have to
+                // recursively search on them
+                if let Some(result) = find_entry_in_dir(&entry.path(), target) {
+                    return Some(result);
+                }
+            }
+        }
+    }
+    None
+}
+
+pub fn parse(entry: impl AsRef<Path>) -> Option<FreeDesktopEntry> {
+    parse_entry(entry.as_ref()).ok()
+}
+
+/// Returns the first FreeDesktop XDG .desktop entry, found inside `dir_path`, when the
+/// "Name" atribute of that entry is `app_name`.
+///
+/// The return value is actually a tuple containing the entry itself, and the path at which
+/// it was found.
+pub fn find_entry_by_app_name(
+    dir_path: &Path,
+    app_name: &OsStr,
+) -> Option<(FreeDesktopEntry, PathBuf)> {
+    for entry in dir_path.read_dir().ok()? {
+        if let Ok(entry) = entry {
+            // If it is a file we open it
+            if entry.path().is_file() {
+                if let Ok(parsed) = parse_entry(entry.path()) {
+                    if let Some(name) = parsed.section("Desktop Entry").attr("Name") {
```

Comment:
```suggestion
                    if parsed
                        .section("Desktop Entry")
                        .attr("Name")
                        .map(str::as_ref)
                        == Some(app_name)
                    {
```
This would save us another level of indentation by letting you remove `if name == app_name` below, but feel free to tell me if you think this looks worse!